### ISSUE TITLE
Allow to add dynamic context to log events

### ISSUE THREAD

**Issue by aethr** [DataDog/browser-sdk#896]
This is a feature request to allow setting context that can be evaluated at the time a log or error occurs.

We use FullStory to record user sessions for UX analysis. The FullStory js library provides a method `FullStory.getCurrentSessionURL(true);` which returns a URL to the recording of the current session *at the current timestamp*. If we could call this method each time an error occurs and send the result along with error or log context, we would be able to go straight from our log entry to a recording of what the user was doing when the error occurred.

Our current approach is to wrap the DataDog Logger in our own logging library where we can call this method in our own `log` method before adding the result to log context and sending it, but this isn't ideal for architectural reasons.

A nice solution to this problem might be adding the ability to attach functions as context, which would be evaluated at the time the log is sent, like:

```js
logger.addContext('static_value', '123');
logger.addContext('fullstory_session_url', (event) => { return FullStory.getCurrentSessionURL(true); });

logger.log('a message', { extra_context: 456 });

// sends context: { static_value: '123', extra_context: 456, fullstory_session_url: 'https://example.com/session/abcd?t=112233' }
```

Alternatively, adding another init option similar to `beforeSend` that allowed extra context to be gathered at the time a log event is created would probably be simple enough to solve our use case.

**Comment by bcaudan** [comment]
Hi @aethr,

That is indeed an interesting feature!
We have recently updated the RUM `beforeSend` API to modify the context of most of the RUM events (cf https://github.com/DataDog/browser-sdk/pull/869, documentation for that should arrive soon), so it is definitely something that we could add for logs too.

We'd be happy to review a PR for it if you want to have a look 😉

**Comment by bcaudan** [comment]
Available in 2.16.0
